--- baresip-4.5.0/include/baresip.h	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/include/baresip-new.h	2026-07-24 18:50:25.823068651 +0400
@@ -292,6 +292,7 @@
 enum sdp_neg_state call_sdp_neg_state(const struct call *call);
 bool call_sdp_change_allowed(const struct call *call);
 
+
 /**
  * Command parameter parse helpers
  */
@@ -1017,7 +1018,6 @@
 bool ua_req_allowed(const struct ua *ua, const struct sip_msg *msg);
 bool ua_req_check_origin(const struct ua *ua, const struct sip_msg *msg);
 
-
 /* One instance */
 int  ua_init(const char *software, bool udp, bool tcp, bool tls);
 void ua_close(void);
@@ -1475,7 +1475,6 @@
 bool audio_txtelev_empty(const struct audio *au);
 void audio_call_telev_handler(const struct audio *au, int key, bool end);
 
-
 /*
  * Video stream
  */
@@ -1915,6 +1914,21 @@
 		    const struct odict *od);
 
 
+/*
+ * empbx extensions
+ */
+void empbx_baresip_uag_set_custom_sip_handler(sip_msg_h *handler);
+
+void empbx_baresip_ua_set_xdata(struct ua *ua, void *data, bool destroy);
+void *empbx_baresip_ua_get_xdata(const struct ua *ua);
+void empbx_baresip_call_set_xdata(struct call *call, void *data, bool destroy);
+void *empbx_baresip_get_xdata(struct call *call);
+void empbx_baresip_audio_set_xdata(struct audio *au, void *data, bool destroy);
+void *empbx_baresip_audio_get_xdata(const struct audio *au);
+void empbx_baresip_video_set_xdata(struct video *vid, void *data, bool destroy);
+void *empbx_baresip_video_get_xdata(const struct video *vid);
+
+
 #ifdef __cplusplus
 }
 #endif
--- baresip-4.5.0/src/core.h	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/core-new.h	2026-02-20 13:53:45.496352551 +0400
@@ -391,6 +391,7 @@
 	uint32_t transports;           /**< Supported transports mask       */
 	bool delayed_close;            /**< Module will close SIP stack     */
 	sip_msg_h *subh;               /**< Subscribe handler               */
+	sip_msg_h *xsiph;              /**< akstel sip extra handler        */
 	ua_exit_h *exith;              /**< UA Exit handler                 */
 	bool nodial;                   /**< Prevent outgoing calls          */
 	void *arg;                     /**< UA Exit handler argument        */
--- baresip-4.5.0/src/uag.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/uag-new.c	2026-07-24 19:02:02.753534953 +0400
@@ -204,15 +204,21 @@
 
 	(void)arg;
 
+    if (uag.xsiph) {
+         if(uag.xsiph(msg, NULL)) {
+            return true;
+        }
+    }
+
 	if (pl_strcmp(&msg->met, "OPTIONS") &&
 	    pl_strcmp(&msg->met, "REFER"))
 		return false;
 
-	ua = uag_find_msg(msg);
-	if (!ua) {
-		(void)sip_treply(NULL, uag_sip(), msg, 404, "Not Found");
-		return true;
-	}
+    ua = uag_find_msg(msg);
+    if (!ua) {
+        (void)sip_treply(NULL, uag_sip(), msg, 404, "Not Found");
+        return true;
+    }
 
 	if (!ua_req_check_origin(ua, msg)) {
 		(void)sip_treply(NULL, uag_sip(), msg, 403, "Forbidden");
@@ -1307,3 +1313,11 @@
 {
 	return uag.subh;
 }
+
+/**
+ * empbx
+ */
+void empbx_baresip_uag_set_custom_sip_handler(sip_msg_h *handler)
+{
+	uag.xsiph = handler;
+}
--- baresip-4.5.0/src/ua.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/ua-new.c	2026-07-24 19:01:30.006734466 +0400
@@ -30,6 +30,8 @@
 	struct list custom_hdrs;     /**< List of outgoing headers           */
 	char *ansval;                /**< SIP auto answer value              */
 	struct sa dst;               /**< Current destination address        */
+	void    *xuser_data;         /** empbx */
+	bool    xud_destroy;         /** empbx */
 };
 
 struct ua_xhdr_filter {
@@ -61,6 +63,10 @@
 	mem_deref(ua->ansval);
 	mem_deref(ua->acc);
 
+	if(ua->xud_destroy) {
+    	    mem_deref(ua->xuser_data);
+	}
+
 	if (uag_delayed_close() && list_isempty(uag_list())) {
 		sip_close(uag_sip(), false);
 	}
@@ -2355,3 +2361,15 @@
 
 	return false;
 }
+
+/* empbx */
+void empbx_baresip_ua_set_xdata(struct ua *ua, void *data, bool destroy) {
+	if (!ua) return;
+    ua->xuser_data = data;
+    ua->xud_destroy = destroy;
+}
+
+void *empbx_baresip_ua_get_xdata(struct ua *ua) {
+    return ua ? ua->xuser_data : NULL;
+}
+
--- baresip-4.5.0/src/call.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/call-new.c	2026-07-24 18:56:33.381600861 +0400
@@ -84,6 +84,8 @@
 	bool use_video;
 	bool use_rtp;
 	struct pl *user_data;      /**< User data related to the call       */
+	void    *xuser_data;       /** empbx */
+	bool    xud_destroy;       /** empbx */
 };
 
 
@@ -363,6 +365,10 @@
 	mem_deref(call->acc);
 	mem_deref(call->user_data);
 
+	if(call->xud_destroy) {
+    	    mem_deref(call->xuser_data);
+	}
+
 	list_flush(&call->custom_hdrs);
 }
 
@@ -3454,3 +3460,15 @@
 		|| (sdp_state == SDP_NEG_NONE
 		|| sdp_state == SDP_NEG_REMOTE_OFFER);
 }
+
+/* empbx */
+void empbx_baresip_call_set_xdata(struct call *call, void *data, bool destroy) {
+    if (!call) return;
+
+    call->xuser_data = data;
+    call->xud_destroy = destroy;
+}
+
+void *empbx_baresip_call_get_xdata(struct call *call) {
+    return call ? call->xuser_data : NULL;
+}
--- baresip-4.5.0/src/audio.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/audio-new.c	2026-07-24 18:54:29.798128742 +0400
@@ -133,6 +133,8 @@
 	audio_level_h *levelh;        /**< Audio level handler             */
 	audio_err_h *errh;            /**< Audio error handler             */
 	void *arg;                    /**< Handler argument                */
+	void *xuser_data;             /** empbx */
+	bool xud_destroy;             /** empbx */
 };
 
 
@@ -214,6 +216,10 @@
 	mem_deref(a->telev);
 	mem_deref(a->aur);
 
+	if(a->xud_destroy) {
+    	    mem_deref(a->xuser_data);
+	}
+
 	mem_deref(a->tx.mtx);
 }
 
@@ -1932,3 +1938,14 @@
 {
 	return au ? &au->cfg : NULL;
 }
+
+/* empbx */
+void empbx_baresip_audio_set_xdata(struct audio *au, void *data, bool destroy) {
+    if (!au) return;
+
+    au->xuser_data = data;
+    au->xud_destroy = destroy;
+}
+void *empbx_baresip_audio_get_xdata(struct audio *au) {
+    return au ? au->xuser_data : NULL;
+}
--- baresip-4.5.0/src/video.c	2026-01-27 18:41:34.000000000 +0400
+++ baresip-4.5.0/src/video-new.c	2026-07-24 18:55:11.956584098 +0400
@@ -169,6 +169,8 @@
 	bool nack_pli;          /**< Send NACK/PLI to peer                */
 	video_err_h *errh;      /**< Error handler                        */
 	void *arg;              /**< Error handler argument               */
+	void *xuser_data;       /** empbx */
+	bool xud_destroy;       /** empbx */
 };
 
 
@@ -310,6 +312,10 @@
 	mtx_unlock(&vrx->lock);
 	mtx_destroy(&vrx->lock);
 
+	if(v->xud_destroy) {
+    	    mem_deref(v->xuser_data);
+	}
+
 	tmr_cancel(&v->tmr);
 	mem_deref(v->strm);
 	mem_deref(v->peer);
@@ -1965,3 +1971,15 @@
 	vid->vtx.picup = true;
 	mtx_unlock(vid->vtx.lock_enc);
 }
+
+/* empbx */
+void empbx_baresip_video_set_xdata(struct video *vid, void *data, bool destroy) {
+    if (!vid) return;
+
+    vid->xuser_data = data;
+    vid->xud_destroy = destroy;
+}
+
+void *empbx_baresip_video_get_xdata(struct video *vid) {
+    return vid ? vid->xuser_data : NULL;
+}
